fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/keyspace/ResolverMappingReplicator.java

```diff
@@ -135,30 +135,22 @@ public void copyTo(LocatableResolver replica) {
                         .setContinuation(continuation.get())
                         .build();
 
-                return AsyncUtil.whileTrue(() ->
-                    cursor.onHasNext().thenCompose(hasNext -> {
-                        CompletableFuture<Void> currentStep;
-                        if (hasNext) {
-                            final KeyValue kv = cursor.next();
-                            final String mappedString = primaryMappingSubspace.unpack(kv.getKey()).getString(0);
-                            final ResolverResult mappedValue = valueDeserializer.apply(kv.getValue());
-                            accumulator.accumulate(mappedValue.getValue());
-                            counter.incrementAndGet();
-
-                            currentStep = replica.setMapping(context, mappedString, mappedValue);
-                        } else {
-                            currentStep = AsyncUtil.DONE;
-                        }
-                        return currentStep.thenApply(ignore2 -> hasNext);
-                    }), context.getExecutor())
-                    .thenCompose(ignore -> context.commitAsync().thenRun(() -> {
-                        byte[] nextContinuation = cursor.getContinuation();
+                return cursor.mapPipelined(kv -> {
+                    final String mappedString = primaryMappingSubspace.unpack(kv.getKey()).getString(0);
+                    final ResolverResult mappedValue = valueDeserializer.apply(kv.getValue());
+                    accumulator.accumulate(mappedValue.getValue());
+                    counter.incrementAndGet();
+                    return replica.setMapping(context, mappedString, mappedValue);
+                }, 1).forEachResult(ignore -> { }).thenCompose(lastResult -> context.commitAsync().thenRun(() -> {
```

Comment:
I took a shot at that. The issue I ran into when I tried that the first time was that we don't have a `map` method that takes a lambda which takes a `RecordCursorResult`, and doing so would involve tinkering with the `MapPipelinedCursor` in ways I didn't totally feel comfortable with. So it was either make a `forEachResultAsync` that returned a `CompletableFuture<RecordCursorResult<T>>` but whose lambda provided only a `T` instead of a `RecordCursorResult<T>` (unlike `forEachResult`), or I could make a `forEachResult` that took a result but didn't allow for pipelining and reimplemented some of the `forEach` logic but more asynchronously.

In this next commit, I went with the latter option, but I'm not totally sure it's right.

Reply:
What you have looks fine to me.